# Basic Problems: (1 - 10)

# Problem 1: (Variable Swap)
Write a Python program to swap the values of two variables without using a temporary variable.

In [ ]:
def swap_vars(a, b):
    print(f"Before swapping: a = {a}, b = {b}")
    a,b = b,a
    print(f"After swapping: a = {a}, b = {b}")

var1 = 5
var2 = 10
swap_vars(var1, var2)


# Problem 2:(Even or Odd)
Write a program that determines if a number is even or odd.

In [2]:
num = int(input('Enter a number: '))
if num % 2==0:
    print(f'The Number {num} is Even.')
else:
    print( f'The Number {num} is odd.')


Enter a number: 5
The Number 5 is odd.


In [3]:
def Even_Odd(a):
    if a % 2 ==0:
        return "Even"
    else:
        return "Odd"

num = int(input('Enter a number: '))
Check_num = Even_Odd(num)
print(f"The number {num} is {Check_num}.")

Enter a number: 6
The number 6 is Even.


# Problem 3: (String Reverse) 
Write a Python function to reverse a given string and return the reversed string.

In [14]:
def reverse_string(input_string):
    return input_string[::-1]   #using slicing

input_str = "STUDYMART!"
reversed_str = reverse_string(input_str)
print("Original string:", input_str)
print("Reversed string:", reversed_str)

Original string: STUDYMART!
Reversed string: !TRAMYDUTS


# Problem 4: (Type Conversion)
Given a list of integers, write a Python program to convert each element of the list to a string.

In [17]:
integer_list = [1, 2, 3, 4, 5]

string_list = [str(i) for i in int_list] ## Using list comprehension

print("List of integers:", integer_list)
print("List of strings:", string_list)

List of integers: [1, 2, 3, 4, 5]
List of strings: ['1', '2', '3', '4', '5']


In [19]:
integer_list = [1, 2, 3, 4, 5]
string_list = []

for num in int_list:
    string_list.append(str(num))

print("List of integers:", integer_list)
print("List of strings:", string_list)

List of integers: [1, 2, 3, 4, 5]
List of strings: ['1', '2', '3', '4', '5']


# Problem 5: Temperature Converter
Write a Python program that converts a temperature in Celsius to Fahrenheit.
Take the Celsius temperature as input from the user.

In [4]:
#Celsius to Fahrenheit

C = float(input('Celsius_Temperature: '))

F = (9*C+160)/5

print('Fahr_Temperature: ',F)

Celsius_Temperature: 50
Fahr_Temperature:  122.0


In [5]:
#Fahrenheit to Celsius

def Fahr_to_Cel(F):
    C = (F-32)*(5/9)
    return C

Farhenheit_Tem = int(input('Farhenheit_Tem: '))
Celsius_Tem = Fahr_to_Cel(Farhenheit_Tem)
print(f"{Farhenheit_Tem}F is equal to {Celsius_Tem}C.")

Farhenheit_Tem: 25
25F is equal to -3.8888888888888893C.


# Problem 6: (Data Type Checker)
Write a Python function that takes a variable as input and returns the data type of the variable as a string (e.g., “int”, “float”, “str”, “list”, etc.).

In [21]:
def data_type(variable):
    dt = type(variable)                 #for example a variable type is <class 'int'> 
    dt_final = str(dt).split("'")[1]  #convert string and after split it turn into a list ["<class ", "int", ">"] 
    return dt_final                    #and [1] is index of "int".

var1 = 10
var2 = 3.14
var3 = "Boltu"
var4 = [1, 2, 3]
var5 = {'a': 1, 'b': 2}


type_var1 = data_type(var1)
type_var2 = data_type(var2)
type_var3 = data_type(var3)
type_var4 = data_type(var4)
type_var5 = data_type(var5)


print(f"Data type of var1: {type_var1}")
print(f"Data type of var2: {type_var2}")
print(f"Data type of var3: {type_var3}")
print(f"Data type of var4: {type_var4}")
print(f"Data type of var5: {type_var5}")


Data type of var1: int
Data type of var2: float
Data type of var3: str
Data type of var4: list
Data type of var5: dict


# Problem 7: (String Palindrome)
Write a program that checks if a given string is a palindrome.

In [4]:
def is_palindrome(s):
    s = s.replace(" ", "").lower()
    left, right = 0, len(s) - 1
    
    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
    
    return True


string = input("Enter a string to check: ")

if is_palindrome(string):
    print(f"The string '{string.upper()}' is a palindrome.")
else:
    print(f"The string '{string.upper()}' is not a palindrome.")


Enter a string to check: madam
The string 'MADAM' is a palindrome.


In [6]:
s = input('Enter a string to check: ')
s = s.replace(" ", "").lower()

start = 0
end = len(s)-1
flag = True
while start < end:
    if s[start] != s[end]:
        flag = False
    start+=1
    end-=1

if flag:
    print(f'Given string {s.upper()} is a palindrome.')
else:
    print(f'Given string {s.upper()} is not a palindrome.')

Enter a string to check: wow
Given string WOW is a palindrome.


# Problem 8: (String Reversal with Slicing)
Write a Python function to reverse a given string using slicing.

In [8]:
def rev_str(input_str):
    reversed_str = input_str[-1::-1]
    return reversed_str

input_str = input('Enter a string: ')
output = rev_str(input_str)
print(f'Reversed of {input_str} is {output}')

Enter a string: PICCKEY
Reversed of PICCKEY is YEKCCIP


# Problem 9: (String Concatenation)
Write a Python program that takes two strings as input and concatenates them into a single string without using the `+` operator.

In [9]:
def concatenate_strings(s1, s2):
    concatenated_string = ''.join([s1, s2]) # Also use: "{}{}".format(s1,s2) or "{0}{1}".format(s1,s2) or "%s%s" % (str1, str2)
    return concatenated_string

str1 = input("Enter the first string: ")
str2 = input("Enter the second string: ")

result = concatenate_strings(str1, str2)
print("Concatenated String:", result)

Enter the first string: ma
Enter the second string: ma
Concatenated String: mama


# Problem 10: (Typecasting Challenge)
Given three variables: `a = ‘100’`, `b = 25`, and `c = ‘10.5’`, write a Python program to perform the following operations and print the results: – Convert `a` to an integer and add it to `b`. – Convert `c` to a float and subtract it from the result of the first operation. – Convert the final result to a string and concatenate it with the string ” is the answer.”

In [2]:
a = '100'
b = 25
c = '10.5'

r1 = int(a) + b

r2 = r1 - float(c)

n = ' is the answer'
final = str(r2) + n 
final

'114.5 is the answer'

# Python Conditional Statements Problems: (11 - 19)

# Problem 11: (Positive, Negative, or Zero)
Write a program that determines if a number is positive, negative, or zero.

In [7]:
x = int(input('Enter a number: '))

if x>0:
    print(f'The number {x} is Positive.')
    
elif x<0:
        print(f'The number {x} is Negative.')
        
else:
    print(f'The number {x} is Zero.')

Enter a number: 25
The number 25 is Positive.


# Problem 12: (Largest of Three Numbers)
Write a program that finds the maximum of three numbers.

In [8]:
def Maximum(x):
    a = max(x)
    return a

nums = [2,4,4]
Max = Maximum(nums)
print(f'Maximum number is {Max}.')


Maximum number is 4.


In [9]:
def Maximum(x1,x2,x3):
    a = max(x1,x2,x3)
    return a

num1 = int(input('1st Number: '))
num2 = int(input('2nd Number: '))
num3 = int(input('3rd Number: ')
           
Max = Maximum(num1,num2,num3)
           
print(f'Maximum number of {num1}, {num2} and {num3} is {Max}.')

1st Number: 23
2nd Number: 10
3rd Number: 25
Maximum number of 23, 10 and 25 is 25.


# Problem 13: (Leap Year checker)
Write a program that determines if a year is a leap year or not.

In [3]:
def is_leap_year(year):
    # Leap years are divisible by 4
    # However, years divisible by 100 are not leap years, unless they are also divisible by 400
    if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
        return True
    else:
        return False

year1 = 2020
year2 = 2021
year3 = 2000
year4 = 1900

print(f"Is {year1} a leap year? {is_leap_year(year1)}")
print(f"Is {year2} a leap year? {is_leap_year(year2)}")
print(f"Is {year3} a leap year? {is_leap_year(year3)}")
print(f"Is {year4} a leap year? {is_leap_year(year4)}")


Is 2020 a leap year? True
Is 2021 a leap year? False
Is 2000 a leap year? True
Is 1900 a leap year? False


# Problem 14: (Grades Classification)
Write a Python program that takes a student’s percentage as input and prints their corresponding grade according to the following criteria: – 90% or above: A+ – 80-89%: A – 70-79%: B – 60-69%: C – Below 60%: Fail

In [11]:
x = int(input('Enter the Percentage: '))
if x>=90:
    print(f'The grade for {x}% is A+.')
elif x>=80:
     print(f'The grade for {x}% is A.')
elif x>=70:
     print(f'The grade for {x}% is B.')
elif x>=60:
     print(f'The grade for {x}% is C.')
else:
     print(f'The grade for {x}% is F.')
    

Enter the Percentage: 52
The grade for 52% is F.


In [12]:
def cal_grade(x):
    if x>=90:
        return "A+"
    elif x>=80:
         return "A"
    elif x>=70:
         return "B"
    elif x>=60:
         return "C"
    else:
         return "F"

percentage = int(input('Enter the percentage: '))
grade = cal_grade(percentage)
print(f'The grade for {percentage}% is {grade}.')

Enter the percentage: 15
The grade for 15% is F.


# Problem 15: (Vowel or Consonant)
Write a Python program that takes a single character as input and determines whether it is a vowel or a consonant.

In [4]:
letter = input('Enter a single character: ')

vowel = {'a','e','i','o','u','A','E','I','O','U'}

if letter in vowel:
    print(f'The letter {letter} is a vowel')
    
else:
    print(f'The letter {letter} is a consonant.')

Enter a single character: y
The letter y is a consonant.


In [9]:
vowels = 'aeiouAEIOU'

letter = input('Enter a single alphabet: ')

if letter in vowels:
    print(f"The alphabet '{letter}' is a vowel.")
else:
    print(f"The character '{letter}' is a consonant.")

Enter a single alphabet: t
The character 't' is a consonant.


# Problem 16. (Time Classification)
Write a Python program that takes the time in hours (24-hour format) as input and prints “Good Morning”, “Good Afternoon”, “Good Evening”, or “Good Night” based on the time.

In [10]:
hour = int(input('Enter a time in hour: '))

if 5 <= hour < 12:
    print("Good Morning")
elif 12 <= hour < 17:
    print("Good Afternoon")
elif 17 <= hour < 21:
    print("Good Evening")
else:
    print("Good Night")

Enter a time in hour: 7
Good Morning


# Problem 17: (Triangle Type Checker)
Write a Python program that takes three sides of a triangle as input and determines whether it forms an equilateral, isosceles, or scalene triangle.

In [3]:
a = int(input('1st side of a triangle: '))
b = int(input('2nd side of a triangle: '))
c = int(input('3rd side of a triangle: '))

if [a,b,c]<=[0,0,0]:
    print("Please enter positive values for sides.")

elif [a+b,b+c,c+a]<=[c,a,b]:
    print('Plz enter 3 valid sides of a triangle.')
        
elif a==b==c:
    print('It is a equilateral triangle.')
    
elif a==b or b==c or c==a:
    print('It is a isosceles triangle.')
    
else:
    print('It is a scalene triangle.')

1st side of a triangle: 5
2nd side of a triangle: 5
3rd side of a triangle: 7
It is a isosceles triangle.


# Problem 18: (Quadratic Equation Solver)
Write a Python program that takes the coefficients (a, b, c) of a quadratic equation as input and calculates and prints the real roots (if they exist) or a message indicating the complex roots.

In [33]:
import cmath

a = float(input("Enter the value of a: "))
b = float(input("Enter the value of b: "))
c = float(input("Enter the value of c: "))

# Calculate the discriminant
d = b**2 - 4*a*c

if d > 0:
    # Two distinct real roots
    r1 = (-b + cmath.sqrt(d)) / (2*a)
    r2 = (-b - cmath.sqrt(d)) / (2*a)
    print(f"The roots are {r1} and {r2}")
elif d == 0:
    # One repeated real root
    r = -b / (2*a)
    print(f"The root is {r}")
else:
    # Two complex roots
    r1 = (-b + cmath.sqrt(d)) / (2*a)
    r2 = (-b - cmath.sqrt(d)) / (2*a)
    print(f"The roots are {r1} and {r2}")


Enter the value of a: 2
Enter the value of b: 3
Enter the value of c: -2
The roots are (0.5+0j) and (-2+0j)


# Problem 19: (Number Ranges)
Write a Python program that takes an integer as input and prints whether the number falls within the ranges: 0-50, 51-100, 101-150, or above 150.

In [2]:
num = int(input('Enter an integer: '))

if 0 <= num <= 50:
    print(f'The number {num} falls within the range 0-50.')

elif 51 <= num <= 100:
    print(f'The number {num} falls within the range 51-100.')
    
elif 101 <= num <= 150:
    print(f'The number {num} falls within the range 101-150.')

else:
    print('The number is above 150.')


Enter an integer: 500
The number is above 150.


# For & While Loops Problems: (20 - 29)

# Problem 20:  (Sum of N Numbers) 
Write a Python program using a for loop to calculate the sum of the first N natural numbers,
where N is taken as input from the user.

In [13]:
#using for loop

N = int(input('Enter the number of N: '))

sum = 0

for i in range(1, N+1):
    sum+=i

print(f'The sum of {N} natural number is {sum}')

Enter the number of N: 23
The sum of 23 natural number is 276


In [14]:
#using while loop

N = int(input('Enter the number of N: '))
sum = 0

while N>0:
    sum+=N
    N-=1

print(f'The sum of {N} natural number is {sum}')

Enter the number of N: 23
The sum of 0 natural number is 276


# Problem 21: (Factorial Calculator)
Write a Python program using a while loop to calculate the factorial of a given number N.

In [15]:
num = int(input('Enter the number:'))

total = 1

if num<0:
        print('The factorial of negative is not defined')

for i in range(num,0,-1):

    total = i * total

print(f'The factorial of {num} is {total}.')

Enter the number:5
The factorial of 5 is 120.


In [16]:
def fact(x):
    
    result = 1
    
    if x==0:
        
        print(f'The factorial of {x} is 1.')
        
    elif x<0:
        print('The factorial of negative is not defined')
        
    else:
        for i in range (x):
            
            result = result*(i+1)
            
        return result

num = int(input('Enter a number: '))
factorial = fact(num)
print(f'The factorial of {num} is {factorial}.')

Enter a number: 5
The factorial of 5 is 120.


In [17]:
def factorial(n):
    if n < 0:
        return "Factorial is not defined for negative numbers."
    elif n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)

num = int(input('Enter a number: '))
result = factorial(num)
print(f"The factorial of {num} is: {result}")


Enter a number: 5
The factorial of 5 is: 120


# Problem 22: (Table of a Number)
Write a Python program using a for loop to print the multiplication table of a given number N.

In [18]:
N = int(input('Multiplication table for: '))
for i in range(1,11):
    print(N,'*',i,'=',N*i)

Multiplication table for: 11
11 * 1 = 11
11 * 2 = 22
11 * 3 = 33
11 * 4 = 44
11 * 5 = 55
11 * 6 = 66
11 * 7 = 77
11 * 8 = 88
11 * 9 = 99
11 * 10 = 110


In [19]:
def mult_table(num):
    print(f"Multiplication table for {num}:")
    for i in range(1, 11):
        print(f"{num} * {i} = {num * i}")


num = int(input("Enter a number: "))

mult_table(num)


Enter a number: 12
Multiplication table for 12:
12 * 1 = 12
12 * 2 = 24
12 * 3 = 36
12 * 4 = 48
12 * 5 = 60
12 * 6 = 72
12 * 7 = 84
12 * 8 = 96
12 * 9 = 108
12 * 10 = 120


# Problem 23: (Count Digits in a Number)
Write a Python program using a while loop to count the number of digits in a given integer N.

In [4]:
def count_digits(n):
    if n == 0:
        return 1
    
    count = 0
    while n != 0:
        n //= 10  # Remove the last digit by integer division
        count += 1

    return count

N = int(input("Enter an integer: "))
result = count_digits(N)
print(f"The number of digits in {N} is: {result}")


Enter an integer: 124
The number of digits in 124 is: 3


In [2]:
N = input('Enter a Integer: ')

num_digit = len(str(N))

print(f"The number of digits in {N} is: {num_digit}")

Enter a Integer: 1254
The number of digits in 1254 is: 4


# Problem 24: (Fibonacci Sequence)
Write a Python program using a for loop to generate the Fibonacci sequence up to a given limit N.

In [19]:
def generate_fibonacci_sequence(n):
    fibonacci_seq = []
    a, b = 0, 1

    if n <= 0:
        return "Please enter a positive integer."
    elif n == 1:
        fibonacci_seq.append(a)
    else:
        fibonacci_seq.extend([a, b])
        for i in range(2, n):
            next_num = a + b
            fibonacci_seq.append(next_num)
            a, b = b, next_num

    return fibonacci_seq

length = int(input("Enter the length of the Fibonacci sequence: "))
result = generate_fibonacci_sequence(length)
print(f"The Fibonacci sequence of length {length} is: {result}")

Enter the length of the Fibonacci sequence: 6
The Fibonacci sequence of length 6 is: [0, 1, 1, 2, 3, 5]


In [20]:
def fibonacci_sequence(n):
    fibonacci = [0, 1]

    if n <= 0:
        return "Please enter a positive integer."
    elif n == 1:
        return [0]
    elif n == 2:
        return fibonacci

    for i in range(2, n):
        next_number = fibonacci[i - 1] + fibonacci[i - 2]
        fibonacci.append(next_number)

    return fibonacci

num = int(input("Enter the length of the Fibonacci sequence: "))
sequence = fibonacci_sequence(num)
print(f"The Fibonacci sequence up to {num} terms is: {sequence}")

Enter the length of the Fibonacci sequence: 6
The Fibonacci sequence up to 6 terms is: [0, 1, 1, 2, 3, 5]


In [2]:
def fibonacci_series(N):
    if N <= 0:
        return "Invalid input. Please provide a positive integer."
    elif N == 1:
        return [0]
    elif N == 2:
        return [0, 1]
    else:
        # Recursive call to generate the rest of the series
        series = fibonacci_series(N - 1)
        series.append(series[-1] + series[-2])
        return series

N = int(input("Enter a positive integer (N) for the Fibonacci series: "))
result = fibonacci_series(N)

print(f"The Fibonacci series up to the {N}th term is: {result}")


Enter a positive integer (N) for the Fibonacci series: 10
The Fibonacci series up to the 10th term is: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


# Problem 25: (Sum of Even Numbers)
Write a Python program using a while loop to calculate the sum of all even numbers between 1 and N, where N is taken as input from the user.

In [23]:
def sum_of_even(n):
    sum_even = 0
    i = 2
    while i<=n:
        sum_even += i
        i += 2
    return sum_even

N = int(input("Enter a number: "))
Total = sum_of_even(N)
print(f"The sum of even numbers between 1 and {N} is: {Total}")


Enter a number: 29
The sum of even numbers between 1 and 29 is: 210


# Problem 26: (Print Patterns)
Write a Python program using nested for loops to print various patterns, such as a right-angled triangle, an inverted right-angled triangle, and so on.

In [7]:
#right-angled triangle

N = int(input("Enter a number: "))

for i in range(1,N+1):
    for l in range(1,i+1):
        print(i,end=' ')
    print()

Enter a number: 5
1 
2 2 
3 3 3 
4 4 4 4 
5 5 5 5 5 


In [8]:
#right-angled triangle

N = int(input("Enter a number: "))

for i in range(1,N+1):
    for l in range(1,i+1):
        print(l,end=' ')
    print()

Enter a number: 5
1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


In [6]:
##right-angled triangle

N = int(input("Enter a number: "))

for i in range(1,N+1):
    for l in range(1,i+1):
        print('*',' ',end='')
    print()

Enter a number: 5
*  
*  *  
*  *  *  
*  *  *  *  
*  *  *  *  *  


In [16]:
#inverted right-angled triangle

N = int(input("Enter a number: "))

for i in range(N,0,-1):
    for l in range(1,i+1):
        print('*',' ',end='')
    print()

Enter a number: 5
*  *  *  *  *  
*  *  *  *  
*  *  *  
*  *  
*  


In [1]:
#Pyramid of star

rows = int(input('Enter the row of pyramid: '))

for i in range (1,rows+1):
    for sapce in range(1,rows-i+1):
        print(end=' ')
    for star in range(1,i+1):
        print('*',end =' ')
    print()

Enter the row of pyramid: 7
      * 
     * * 
    * * * 
   * * * * 
  * * * * * 
 * * * * * * 
* * * * * * * 


In [9]:
#Solid Pyramid of star

rows = int(input('Enter the row of pyramid: '))

for i in range (1,rows+1):
    for sapce in range(1,rows-i+1):
        print(end=' ')
    for star in range(1,2*i):
        print('*',end ='')
    print()

Enter the row of pyramid: 5
    *
   ***
  *****
 *******
*********


In [2]:
#number pyramid

rows = int(input('Enter the row of pyramid: '))

for i in range (1,rows+1):
    for sapce in range(1,rows-i+1):
        print(end=' ')
    for star in range(1,i+1):
        print(i,end =' ')
    print()

Enter the row of pyramid: 5
    1 
   2 2 
  3 3 3 
 4 4 4 4 
5 5 5 5 5 


In [3]:
#Reverse Pyramid

rows = int(input("Enter a number: "))

for i in range(rows,0,-1):
    
    for space in range(0,rows-i):
        print(end=' ')
    
    for l in range(1,i+1):
        
        print('*',end=' ')
        
    print()

Enter a number: 5
* * * * * 
 * * * * 
  * * * 
   * * 
    * 


In [4]:
rows = int(input("Enter a number: "))

for i in range(rows,0,-1):
    
    for space in range(0,2*(rows-i)):
        print(end=' ')
    
    for l in range(1,i+1):
        
        print('*',end=' ')
    print()

Enter a number: 6
* * * * * * 
  * * * * * 
    * * * * 
      * * * 
        * * 
          * 


In [4]:
rows = int(input("Enter a number: "))

for i in range(1,rows+1):
    
    for space in range(0,2*(rows-i)):
        print(end=' ')
    
    for l in range(1,i+1):
        
        print('*',end=' ')
    print()

Enter a number: 5
        * 
      * * 
    * * * 
  * * * * 
* * * * * 


In [5]:
rows = int(input("Enter a number: "))

for i in range(rows,0,-1):
    
    for space in range(0,2*(rows-i)):
        print(end=' ')
    
    for l in range(1,i+1):
        
        print('*',end=' ')
    print()
    
for i in range(1,rows+1):
    
    for space in range(0,2*(rows-i)):
        print(end=' ')
    
    for l in range(1,i+1):
        
        print('*',end=' ')
    print()

Enter a number: 5
* * * * * 
  * * * * 
    * * * 
      * * 
        * 
        * 
      * * 
    * * * 
  * * * * 
* * * * * 


In [5]:
#Sandglass Pattern of Stars

rows = int(input("Enter a number: "))

for i in range(rows,0,-1):
    
    for space in range(0,rows-i):
        print(end=' ')
    
    for l in range(1,i+1):
        print('*',end=' ')
    print()
for i in range (1,rows+1):
    for sapce in range(1,rows-i+1):
        print(end=' ')
    for star in range(1,i+1):
        print('*',end =' ')
    print()

Enter a number: 5
* * * * * 
 * * * * 
  * * * 
   * * 
    * 
    * 
   * * 
  * * * 
 * * * * 
* * * * * 


In [3]:
#Right Pyramid of Stars

N = int(input("Enter a number: "))

for i in range(1,N):
    for l in range(1,i+1):
        print('*',end=' ')
    print()

for i in range(N,0,-1):
    for l in range(1,i+1):
        print('*',end=' ')
    print()

Enter a number: 5
* 
* * 
* * * 
* * * * 
* * * * * 
* * * * 
* * * 
* * 
* 


# Problem 27: (Prime Number Checker)
Write a Python program using a while loop to check if a given number N is prime or not.

In [24]:
def is_prime(number):
    if number <= 1:
        return False
    for i in range(2, int(number ** 0.5) + 1):
        if number % i == 0:
            return False
    return True


num = int(input("Enter a number: "))
if is_prime(num):
    print(f"{num} is a prime number.")
else:
    print(f"{num} is not a prime number.")


Enter a number: 23
23 is a prime number.


In [25]:
N = int(input('Enter the integer: '))

if N <= 1:
    print(f'{N} is not a prime number.')
    
elif N>1:
    for i in range(2,N):
       if N % i ==0:
        print(f'{N} is not a prime number.')
        print(f'{i} is a factor of {N}.')
        break
    else:
        print(f'{N} is a prime number.')
    

Enter the integer: 23
23 is a prime number.


In [26]:
#Some of prime numbers between two numbers.

n1 = int(input('Enter lower number: '))
n2 = int(input('Enter upper number: '))

sum = 0

for i in range(n1,n2):
    
    for l in range(2,i):
       if i % l == 0:
        break
    else:
        print(i)
        sum+=i
        
print(f'Sum of prime numbers between {n1} to {n2} is {sum}.')

Enter lower number: 20
Enter upper number: 40
23
29
31
37
Sum of prime numbers between 20 to 40 is 120.


# Problem 28: (List Manipulation)
Given a list of integers, write a Python program using a for loop to find the sum, average, maximum, and minimum values in the list.

In [5]:
numbers = [1, 35, 54, -99, 67, 2, 9]

total = 0
average = 0
maximum = numbers[0]
minimum = numbers[0]

for num in numbers:
    total += num
    
    if num > maximum:
        maximum = num
    if num < minimum:
        minimum = num

average = total / len(numbers)

print("The sum of the list is", total)
print("The average of the list is", average)
print("The maximum of the list is", maximum)
print("The minimum of the list is", minimum)


The sum of the list is 69
The average of the list is 9.857142857142858
The maximum of the list is 67
The minimum of the list is -99


# Problem 29: (Reverse String)
Write a Python program using a while loop to reverse a given string.

In [46]:
#Using for loop

rev_st = 'STUDYMART'

reversed_string = ''
    
for i in range(len(rev_st)-1,-1,-1):
    reversed_string += rev_st[i]
        
print(reversed_string)

TRAMYDUTS


In [39]:
#Using while loop

def rev_st(input_string):
    reversed_string = ''
    index = len(input_string)-1
    
    while index>=0:
        reversed_string += input_string[index]
        index-=1
    return reversed_string

input_st = input('Enter a string to be reversed: ')
output = rev_st(input_st)
print(output)

Enter a string to be reversed: STUDYMART
TRAMYDUTS


In [41]:
#using inbuilt 'reversed' function

input_str = 'STUDY'

rev_string = reversed(input_str)

print(''.join(rev_string))

YDUTS


In [45]:
#using extended Slicing

input_str = input('Enter a string: ')

reversed_str = input_str[-1::-1]

print(f'Reversed of {input_str} is {reversed_str}')

Enter a string: LOVE
Reversed of LOVE is EVOL


# Nested Loops Problems: (30 - 34)

# Problem 30: (Multiplication Table)
Write a Python program using nested loops to print the multiplication table from 1 to 10.

In [20]:
for i in range(1, 11):
    print(f"Multiplication table for {i}:")
    for j in range(1, 11):
        print(f"{i} * {j} = {i * j}")

Multiplication table for 1:
1 * 1 = 1
1 * 2 = 2
1 * 3 = 3
1 * 4 = 4
1 * 5 = 5
1 * 6 = 6
1 * 7 = 7
1 * 8 = 8
1 * 9 = 9
1 * 10 = 10
Multiplication table for 2:
2 * 1 = 2
2 * 2 = 4
2 * 3 = 6
2 * 4 = 8
2 * 5 = 10
2 * 6 = 12
2 * 7 = 14
2 * 8 = 16
2 * 9 = 18
2 * 10 = 20
Multiplication table for 3:
3 * 1 = 3
3 * 2 = 6
3 * 3 = 9
3 * 4 = 12
3 * 5 = 15
3 * 6 = 18
3 * 7 = 21
3 * 8 = 24
3 * 9 = 27
3 * 10 = 30
Multiplication table for 4:
4 * 1 = 4
4 * 2 = 8
4 * 3 = 12
4 * 4 = 16
4 * 5 = 20
4 * 6 = 24
4 * 7 = 28
4 * 8 = 32
4 * 9 = 36
4 * 10 = 40
Multiplication table for 5:
5 * 1 = 5
5 * 2 = 10
5 * 3 = 15
5 * 4 = 20
5 * 5 = 25
5 * 6 = 30
5 * 7 = 35
5 * 8 = 40
5 * 9 = 45
5 * 10 = 50
Multiplication table for 6:
6 * 1 = 6
6 * 2 = 12
6 * 3 = 18
6 * 4 = 24
6 * 5 = 30
6 * 6 = 36
6 * 7 = 42
6 * 8 = 48
6 * 9 = 54
6 * 10 = 60
Multiplication table for 7:
7 * 1 = 7
7 * 2 = 14
7 * 3 = 21
7 * 4 = 28
7 * 5 = 35
7 * 6 = 42
7 * 7 = 49
7 * 8 = 56
7 * 9 = 63
7 * 10 = 70
Multiplication table for 8:
8 * 1 = 8
8 * 2 = 16

# Problem 31: Pattern Print

In [17]:
N = 5

for i in range(1,N+1):
    for l in range(1,i+1):
        print('*',end='')
    print()

*
**
***
****
*****


# Problem 32. (Matrix Multiplication)
Write a Python program using nested loops to multiply two matrices.

In [4]:
A = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]

B = [[10, 11, 12],
     [13, 14, 15],
     [16, 17, 18]]

result = [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]]

for i in range(len(A)): # Loop through the rows of A
    for j in range(len(B[0])): # Loop through the columns of B
        for k in range(len(B)): # Loop through the rows of B
            
            result[i][j] += A[i][k] * B[k][j]

for row in result:
    print(row)


[84, 90, 96]
[201, 216, 231]
[318, 342, 366]


# Problem 33: (Chessboard Pattern)
Write a Python program using nested loops to print a chessboard pattern (alternating “X” and “O” characters) of size 8×8.

In [6]:
board_size = 8

for i in range(board_size):

    for j in range(board_size):
        
        if (i + j) % 2 == 0:
            print("X", end=" ")
            
        else:
            print("O", end=" ")
    
    print()


X O X O X O X O 
O X O X O X O X 
X O X O X O X O 
O X O X O X O X 
X O X O X O X O 
O X O X O X O X 
X O X O X O X O 
O X O X O X O X 


# Problem 34: (Number Pyramid)
Write a Python program using nested loops to print a number pyramid like the following: 1 22 333 4444 55555.

In [2]:
#number pyramid

rows = int(input('Enter the row of pyramid: '))

for i in range (1,rows+1):
    for sapce in range(1,rows-i+1):
        print(end=' ')
    for star in range(1,i+1):
        print(i,end =' ')
    print()

Enter the row of pyramid: 5
    1 
   2 2 
  3 3 3 
 4 4 4 4 
5 5 5 5 5 


# List Problems: (34 - 43)

# Problem 34: (List Sum)
Write a Python program to find the sum of all elements in a given list of integers.

In [7]:
list_of_integer = [10,15,18]

sum_element = sum(list_of_integer[::])

print('Sum of elements of a list: ',sum_element)

Sum of element of a list:  43


# Problem 35: (List Average)
Write a Python program to calculate the average of all elements in a given list of integers.

In [9]:
list_of_integer = [10,15,18]

average = sum(list_of_integer[::])/len((list_of_integer[::]))

print('Average of elements of a list: ',average)

Average of elements of a list:  14.333333333333334


# Problem 36: (List Max and Min)
Write a Python program to find the maximum and minimum values in a given list of integers.

In [11]:
list_of_integer = [10,15,18]

max_value = max(list_of_integer[::])

min_value = min(list_of_integer[::])

print('Maximum value: ',max_value)
print('Minimum value: ',min_value)

Maximum value:  18
Minimum value:  10


# Problem 37: (List Sorting)
Write a Python program to sort a list of integers in ascending order.

In [16]:
list_of_integer = [10,15,18,100,0,52]

print ('Sorted list (ascending): ',list_of_integer.sort())

#Why This code [list_of_integer.sort()] doesn't work??

Sorted_list = sorted(list_of_integer)

print ('Sorted list (ascending): ',Sorted_list)


Sorted list (ascending):  None
Sorted list (ascending):  [0, 10, 15, 18, 52, 100]


# Program 38: (List Filtering)
Given a list of integers, write a Python program to create a new list that contains only the even numbers from the original list.

In [20]:
original_list = [10,15,18,100,0,52]

new_list = []

for num in original_list:
    if num%2==0:
        new_list.append(num)

print('List of even numbers: ',new_list)

List of even numbers:  [10, 18, 100, 0, 52]


# Program 39.: (List Reversal)
Write a Python program to reverse a given list without using any built-in functions.

In [29]:
#Using for loop

original_list = [10,15,18,100,0,52]

reversal_list = []

for i in range(len(original_list)-1,-1,-1):
    reversal_list.append(original_list[i])

print('Reversal list: ',reversal_list)

Reversal list:  [52, 0, 100, 18, 15, 10]


In [1]:
#Using While loop

original_list = [10,15,18,100,0,52]

reversal_list = []

index = len(original_list)-1

while index>=0:
    reversal_list.append(original_list[index])
    index-=1
    
print('Reversal list: ',reversal_list)

Reversal list:  [52, 0, 100, 18, 15, 10]


In [2]:
#extra for understanding

original_list = [10,15,18,100,0,52]

reversal_list = original_list[::-1]
reversal_list

[52, 0, 100, 18, 15, 10]

# Problem 40: (List Manipulation)
Given two lists of integers, write a Python program to create a new list that contains elements common to both lists.

In [4]:
list_1 = [12,45,14]
list_2 = [65,14,45,56]

new_list = []

for i in list_1:
    for j in list_2:
        if i==j:
            new_list.append(i)

print('New list that contains common elements: ',new_list)

New list that contains common elements:  [45, 14]


# Problem 41: (List Element Count)
Write a Python program to count the occurrences of a specific element in a given list.

In [2]:
#using count function

my_list = [10, 12, 3, 4, 12, 5, 12, 6, 12]
element_to_count = 12

occur_el = my_list.count(element_to_count)

print(f"The element {element_to_count} occurs {occur_el} times in the list.")


The element 12 occurs 4 times in the list.


In [7]:
my_list = [10, 12, 3, 4, 12, 5, 12, 6, 12]
element_to_count = 12

count = 0
for item in my_list:
    if item==element_to_count:
        count+=1
        
print(f"The element {element_to_count} occurs {count} times in the list.")

The element 12 occurs 4 times in the list.


In [8]:
def count_occur(My_list, element):
    
    count = 0
    for item in My_list:
        if item == element:
            count += 1
    return count

my_list = [10, 12, 3, 4, 12, 5, 12, 6, 12]
element_to_count = 12

occur_el = count_occur(my_list, element_to_count)
print(f"The element {element_to_count} occurs {occur_el} times in the list.")


The element 12 occurs 4 times in the list.


# Problem 42: (List Duplicates Removal)
Write a Python program to remove duplicates from a given list while preserving the order of the elements.

In [9]:
#Using for loop

def remove_duplicates(my_list):
    
    new_list = []
    for item in my_list:
        if item not in new_list:
            new_list.append(item)
    return new_list

original_list = [1, 2, 3, 2, 4, 1, 5, 6, 4]

result_list = remove_duplicates(original_list)
print(f"Original List: {original_list}")
print(f"List without Duplicates: {result_list}")

Original List: [1, 2, 3, 2, 4, 1, 5, 6, 4]
List without Duplicates: [1, 2, 3, 4, 5, 6]


In [13]:
#using while loop

my_list = [1, 2, 3, 2, 4, 1, 5, 6, 4]

new_list = []
index = 0

while index < len(my_list):
    item = my_list[index]
    if item not in new_list:
        new_list.append(item)
    index += 1

print(f"Original List: {my_list}")
print(f"List with Duplicates Removed: {new_list}")


Original List: [1, 2, 3, 2, 4, 1, 5, 6, 4]
List with Duplicates Removed: [1, 2, 3, 4, 5, 6]


# Problem 43: (List Comprehension)
Given a list of integers, write a Python program to create a new list that contains the squares of the elements using list comprehension.

In [15]:
my_list = [10, 12, 3, 4, 12, 5, 12, 6, 12]

modified_list = [item*item for item in my_list]

print(f"Original List: {my_list}")
print(f"Modified List: {modified_list}")

Original List: [10, 12, 3, 4, 12, 5, 12, 6, 12]
Modified List: [100, 144, 9, 16, 144, 25, 144, 36, 144]


# Nested List Problems: (44 - 48)

# Problem 44: (Matrix Addition)
Write a Python program to add two matrices represented as nested lists.

In [7]:
X = [ [1, 2, 3], [4, 5, 6], [7, 8, 9] ]
Y = [ [9, 8, 7], [6, 5, 4], [3, 2, 1] ]

result = []

for i in range(len(X)):
    row = []
    
    for j in range(len(X[0])):
       
        row.append(X[i][j] + Y[i][j])
    
    result.append(row)

print(result)


[[10, 10, 10], [10, 10, 10], [10, 10, 10]]


# Problem 45: (Flatten Nested List)
Write a Python program to flatten a given nested list and convert it into a single-dimensional list.

In [8]:
def f_list(n_list):
    single_d_list = []
    
    for item in n_list:
        if isinstance(item, list):
            single_d_list.extend(f_list(item))
            
        else:
            single_d_list.append(item)
            
    return single_d_list

nested_list = [1, [2, [3, 4, [5, 6]], 7], 8, [9, 10]]

single_dimensional_list = f_list(nested_list)

print("Original Nested List:", nested_list)
print("Flattened List:",single_dimensional_list)

Original Nested List: [1, [2, [3, 4, [5, 6]], 7], 8, [9, 10]]
Flattened List: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


# Problem 46: (List Element Frequency)
Given a nested list containing lists of integers, write a Python program to count the frequency of each element in the entire nested list.

In [10]:
from collections import Counter

nested_list = [[1, 2, 3], [2, 3, 4], [3, 4, 5], [4, 5, 6]]

flattened_list = [item for sublist in nested_list for item in sublist]

element_counts = Counter(flattened_list)

print("Nested List:", nested_list)
print("Element Frequency Counts:", element_counts)

Nested List: [[1, 2, 3], [2, 3, 4], [3, 4, 5], [4, 5, 6]]
Element Frequency Counts: Counter({3: 3, 4: 3, 2: 2, 5: 2, 1: 1, 6: 1})


# Problem 47: (Transpose Matrix)
Write a Python program to transpose a given matrix represented as a nested list.

In [11]:
matrix = [ [1, 2, 3], [4, 5, 6], [7, 8, 9] ]

t_matrix = []

for i in range(len(matrix[0])):
    row = []
   
    for j in range(len(matrix)):
        
        row.append(matrix[j][i])
    
    t_matrix.append(row)

for row in t_matrix:
    print(row)

[1, 4, 7]
[2, 5, 8]
[3, 6, 9]


In [12]:
original_matrix = [[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]]

transposed_matrix = list(map(list, zip(*matrix)))

print("Original Matrix:")
for row in original_matrix:
    print(row)

print("\nTransposed Matrix:")
for row in transposed_matrix:
    print(row)

Original Matrix:
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]

Transposed Matrix:
[1, 4, 7]
[2, 5, 8]
[3, 6, 9]


# Problem 48: (List of Lists Concatenation)
Given a list of nested lists, write a Python program to concatenate all the sublists into a single flat list.

In [13]:
nested_list = [ [1, 2, 3], [4, 6], [7, 8, 9] ]

flat_list = []

for sublist in nested_list:
    
    for element in sublist:

        flat_list.append(element)

print(flat_list)

[1, 2, 3, 4, 6, 7, 8, 9]


# Tuple Problems: (49 - 58)

# Problem 49: (Tuple Concatenation)
Write a Python program to concatenate two tuples and create a new tuple.

In [16]:
tuple1 = ('a','b','c')
tuple2 = ('d','c','f')

new_tuple = tuple1 + tuple2
new_tuple

('a', 'b', 'c', 'd', 'c', 'f')

# Problem 50: (Tuple Unpacking)
Given a tuple with three elements (x, y, z), write a Python program to unpack the tuple and assign the values to three variables.

In [18]:
my_tup = ('x','y','z')

var1,var2,var3 = my_tup

print('variable_1: ',var1)
print('variable_2: ',var2)
print('variable_3: ',var3)

variable_1:  x
variable_2:  y
variable_3:  z


# Problem 51: (Tuple Sorting)
Write a Python program to sort a tuple of integers in ascending order.

In [19]:
my_tup = (10,15,17,2,84,23,15,5,7,2)

sorted_tuple = tuple(sorted(my_tup)) #sorting list: new_list = sorted(my_list)

sorted_tuple

(2, 2, 5, 7, 10, 15, 15, 17, 23, 84)

# Problem 52: (Tuple Frequency Count)
Given a tuple containing various elements, write a Python program to count the frequency of a specific element in the tuple.

In [20]:
my_tup = (10,15,17,2,15,23,15,5,7,2,'a','e','a')

specific_element = input("Enter an element to count: ")

element_frequency = my_tup.count(specific_element)

print(f"Frequency of {specific_element} in the tuple: {element_frequency}")

Enter an element to count: 2
Frequency of 2 in the tuple: 0


# Problem 53: (Tuple to List)
Write a Python program to convert a tuple into a list.

In [22]:
my_tuple = (2,'e',5,'g')

new_list = list(my_tup)

print("Original Tuple:", my_tuple)
print("Converted List:", new_list)

Original Tuple: (2, 'e', 5, 'g')
Converted List: [2, 'e', 5, 'g']


# Problem 54: (Tuple Reversal)
Write a Python program to reverse a tuple without using any built-in functions.

In [28]:
my_tuple = (2,'e',5,'g')

reverse_tuple = ()

for i in range(len(my_tuple)-1,-1,-1):
    reverse_tuple = reverse_tuple + tuple((my_tuple[i],)) 
    
#For tuple with single item: my_tup = (item,)
#here last comma(,) is mandatory, otherwise it int.

    
print('Original Tuple: ',my_tuple)
print('Reversed Tuple: ',reverse_tuple)
    

Original Tuple:  (2, 'e', 5, 'g')
Reversed Tuple:  ('g', 5, 'e', 2)


# Problem 55: (Tuple Slicing)
Given a tuple, write a Python program to extract a slice of elements from it.

In [3]:
my_tuple = (1, 2, 3, 4, 5, 6, 7, 8, 9)

slice1 = my_tuple[2:5] 
slice2 = my_tuple[:5]
slice3 = my_tuple[5:]
slice4 = my_tuple[1:8:2]

print("slice 1:", slice1)
print("slice 2:", slice2)
print("slice 3:", slice3)
print("slice 4:", slice4)

slice 1: (3, 4, 5)
slice 2: (1, 2, 3, 4, 5)
slice 3: (6, 7, 8, 9)
slice 4: (2, 4, 6, 8)


# Problem 56: (Tuple Operations)
Given two tuples of integers, write a Python program to perform element-wise addition, subtraction, and multiplication and create new tuples for each operation.

In [3]:
#using for loop

tuple1 = [1, 2, 3]
tuple2 = [4, 5, 6]

addition = []
subtraction = []
multiplication = []

for i in range(len(tuple1)):
    addition.append(tuple1[i] + tuple2[i])
    
for i in range(len(tuple1)):
    subtraction.append(tuple1[i] - tuple2[i])
    
for i in range(len(tuple1)):
    multiplication.append(tuple1[i] * tuple2[i])

print('New tupple with Addition: ', addition)
print('New tupple with subtraction: ', subtraction)
print('New tupple with multiplication: ', multiplication)

New tupple with Addition:  [5, 7, 9]
New tupple with subtraction:  [-3, -3, -3]
New tupple with multiplication:  [4, 10, 18]


In [6]:
#Using List Comprehension:

tuple1 = [1, 2, 3]
tuple2 = [4, 5, 6]

addition = [x + y for x, y in zip(tuple1, tuple2)]
subtraction = [x - y for x, y in zip(tuple1, tuple2)]
multiplication= [x * y for x, y in zip(tuple1, tuple2)]

print('New tupple with Addition: ', addition)
print('New tupple with subtraction: ', subtraction)
print('New tupple with multiplication: ', multiplication)

New tupple with Addition:  [5, 7, 9]
New tupple with subtraction:  [-3, -3, -3]
New tupple with multiplication:  [4, 10, 18]


# Problem 57: (Tuple Membership Test)
Write a Python program that takes an element as input and checks if it exists in a given tuple.

In [7]:
my_tuple = (10, 12, 'radin', 4.5, 'mango15
            ', True)

input_element = input("Enter an element to check: ")

if input_element in my_tuple:
    print(f"{input_element} exists in the tuple.")
    
else:
    print(f"{input_element} does not exist in the tuple.")

Enter an element to check: 15
15 does not exist in the tuple.


# Problem 58: (Tuple Packing)
Write a Python program to pack three variables into a single tuple and print the tuple.

In [8]:
var1 = 724
var2 = 'study'
var3 = [10, 25, 3]

packed_tuple = (var1, var2, var3)

print("Packed Tuple:", packed_tuple)


Packed Tuple: (724, 'study', [10, 25, 3])


# Nested List Problems: (59 - 68)

# Problem 59: (Nested List Element Access)
Given a nested list, write a Python program to access and print specific elements from it.

In [12]:
nested_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

element_1_2 = nested_list[0][1]
element_2_1 = nested_list[1][0]

print(f"Element at position (1, 2): {element_1_2}")
print(f"Element at position (2, 1): {element_2_1}")


Element at position (1, 2): 2
Element at position (2, 1): 4


In [10]:
nested_list = [ [1, 2, 3], [4, 5, 6], [7, 8, 9]]

for sublist in nested_list: # Loop over the outer list
    for element in sublist: # Loop over the inner list
        print(element, end=" ") # Print each element with a space
    print() # Print a new line after each sublist

1 2 3 
4 5 6 
7 8 9 


# Problem 60. (Nested List Flattening)
Write a Python program to flatten a nested list and convert it into a single-dimensional list.

In [14]:
def flatten_list(nested_list):
    
    flattened_list = []
    
    for element in nested_list:
        if isinstance(element, list):
            flattened_list.extend(flatten_list(element))
        else:
            flattened_list.append(element)
    return flattened_list

nested_list = [1, [2, 3, [4, 5]], 6, [7, 8, [9, 10]]]

result_list = flatten_list(nested_list)

print("Original Nested List:")
print(nested_list)

print("\nFlattened List:")
print(result_list)


Original Nested List:
[1, [2, 3, [4, 5]], 6, [7, 8, [9, 10]]]

Flattened List:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


# Problem 61: (Nested List Sorting)
Given a nested list containing lists of integers, write a Python program to sort the sublists based on their lengths.

In [15]:
nested_list = [
    [1, 2, 3],
    [4, 5],
    [6, 7, 8, 9],
    [10, 11, 12]
]

sorted_nested_list = sorted(nested_list, key=len)

print("Original Nested List:")
for sublist in nested_list:
    print(sublist)

print("\nSorted Nested List based on Lengths:")
for sublist in sorted_nested_list:
    print(sublist)


Original Nested List:
[1, 2, 3]
[4, 5]
[6, 7, 8, 9]
[10, 11, 12]

Sorted Nested List based on Lengths:
[4, 5]
[1, 2, 3]
[10, 11, 12]
[6, 7, 8, 9]


# Problem 62: (List of Tuples Conversion)
Given a nested list containing tuples of (x, y) coordinates, write a Python program to convert it into a list of x-coordinates and a list of y-coordinates.

In [17]:
nested_list = [(1, 2), (3, 4), (5, 6), (7, 8)]

# Extract x-coordinates and y-coordinates
x_coordinates = [x for x, y in nested_list]
y_coordinates = [y for x, y in nested_list]

print("Original Nested Coordinates:")
for coordinate in nested_list:
    print(coordinate)

print("\nX-Coordinates:")
print(x_coordinates)

print("\nY-Coordinates:")
print(y_coordinates)


Original Nested Coordinates:
(1, 2)
(3, 4)
(5, 6)
(7, 8)

X-Coordinates:
[1, 3, 5, 7]

Y-Coordinates:
[2, 4, 6, 8]


# Poblem 63: (Matrix Transpose)
Write a Python program to transpose a given matrix represented as a nested list.

In [19]:
def transpose_matrix(matrix):
    # Use the zip function to transpose the matrix
    transposed_matrix = list(map(list, zip(*matrix)))
    return transposed_matrix

matrix = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]

transposed_result = transpose_matrix(matrix)

print("Original Matrix:")
for row in matrix:
    print(row)

print("\nTransposed Matrix:")
for row in transposed_result:
    print(row)

Original Matrix:
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]

Transposed Matrix:
[1, 4, 7]
[2, 5, 8]
[3, 6, 9]


In [22]:
matrix = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]

# Use zip to group the corresponding elements of each row
# Convert the resulting tuples to lists using list comprehension
transposed_matrix = [list(row) for row in zip(*matrix)]

print("Original matrix:")
for row in matrix:
    print(row)

print("Transposed matrix:")
for row in transposed_matrix:
    print(row)


Original matrix:
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]
Transposed matrix:
[1, 4, 7]
[2, 5, 8]
[3, 6, 9]


In [21]:
def transpose_matrix(matrix):
    
    transposed_matrix = [[matrix[j][i] for j in range(len(matrix))] for i in range(len(matrix[0]))]
    
    return transposed_matrix

matrix = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]

transposed_result = transpose_matrix(matrix)

print("Original Matrix:")
for row in matrix:
    print(row)

print("\nTransposed Matrix:")
for row in transposed_result:
    print(row)

Original Matrix:
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]

Transposed Matrix:
[1, 4, 7]
[2, 5, 8]
[3, 6, 9]


# Problem 64: (Nested List Concatenation)
Given a list of nested lists, write a Python program to concatenate all the sublists into a single flat list.

In [23]:
nested_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

flat_list = [item for sublist in nested_list for item in sublist]

print("Original list:")
print(nested_list)

print("Flattened list:")
print(flat_list)

Original list:
[[1, 2, 3], [4, 5, 6], [7, 8, 9]]
Flattened list:
[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [25]:
from itertools import chain

nested_list = [[1, 2, 3],[4, 5],[6, 7, 8]]
   
flat_list = list(chain(*nested_list))
# Use itertools.chain to concatenate sublists into a single flat list

print("Nested List:")
for sublist in nested_list:
    print(sublist)

print("\nFlat List:")
print(flat_list)

Nested List:
[1, 2, 3]
[4, 5]
[6, 7, 8]

Flat List:
[1, 2, 3, 4, 5, 6, 7, 8]


# Problem 65: (Count Even Numbers)
Write a Python program to count the number of even numbers in a nested list.

In [26]:
def count_even_numbers(nested_list):
    even_count = 0
    for sublist in nested_list:
        for number in sublist:
            if number % 2 == 0:
                even_count += 1
    return even_count

nested_list = [[1, 2, 3],[4, 5, 6],[7, 8, 9]]

even_count_result = count_even_numbers(nested_list)

print("Nested List:")
for sublist in nested_list:
    print(sublist)

print("\nNumber of Even Numbers:", even_count_result)

Nested List:
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]

Number of Even Numbers: 4


# Problem 66: (Maximum Element in Nested List)
Write a Python program to find the maximum element in a nested list of integers.

In [27]:
def find_max_element(nested_list):
    
    max_element = float('-inf') #For non-negative element we can let [max_element=0]

    for sublist in nested_list:
        for number in sublist:
            if number > max_element:
                max_element = number

    return max_element

nested_list = [ [1, 2, 3],[4, 5, 6],[7, 8, 9]]

max_element_result = find_max_element(nested_list)

print("Nested List:")
for sublist in nested_list:
    print(sublist)

print("\nMaximum Element:", max_element_result)


Nested List:
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]

Maximum Element: 9


# Problem 67: (Diagonal Sum of Matrix)
Given a square matrix represented as a nested list, write a Python program to calculate the sum of the elements in the main diagonal.

In [28]:
matrix = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]

diagonal_sum = 0

for i in range(len(matrix)):
    
    diagonal_sum += matrix[i][i]

print("The sum of the main diagonal elements is:", diagonal_sum)

The sum of the main diagonal elements is: 15


# Problem 68: (Nested List Element Search)
Write a Python program to search for a specific element in a nested list and return its position (row and column indices).

In [29]:
def search_nested_list(nested_list, element):
    # Loop through the rows of the nested list
    for i in range(len(nested_list)):
        # Loop through the columns of the current row
        for j in range(len(nested_list[i])):
            # Check if the element is equal to the current element
            if element == nested_list[i][j]:
                return (i, j)
    
    return None

nested_list = [[1, 2, 3],
               [4, 5, 6],
               [7, 8, 9]]

print(search_nested_list(nested_list, 5))
print(search_nested_list(nested_list, 10))


(1, 1)
None


In [32]:
def find_element(matrix, target):
    for i, row in enumerate(matrix):
        for j, element in enumerate(row):
            if element == target:
                return i, j
    return None 

matrix = [
    [1, 2, 3], 
    [4, 5, 6], 
    [7, 8, 9] ]

target_element = 5

result = find_element(matrix, target_element)

print("Matrix:")
for row in matrix:
    print(row)

if result is not None:
    print("\nElement", target_element, "found at position:", result)
else:
    print("\nElement", target_element, "not found in the matrix.")


Matrix:
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]

Element 5 found at position: (1, 1)


# Set Problems: (69 - 76)

# Problem 69: (Duplicate Removal)
Write a Python program that takes a list of elements as input and creates a new set containing only the unique elements from the list.

In [6]:
def create_set(input_list):
    unique_set = set(input_list) #using set() contractor we can convert list/tuple into a set.
    return unique_set

input_list = input("Enter a list of elements separated by spaces: ").split(' ')
#input_list = input("Enter a list of elements separated by comma: ").split(',')

result_set = create_set(input_list)

print("Original List:", input_list)
print("Unique Set:", result_set)


Enter a list of elements separated by spaces: 12 14 16
Original List: ['12', '14', '16']
Unique Set: {'16', '14', '12'}


# Problem 70:(Set Intersection) 
Given two sets A and B, write a Python program to find their intersection and print the common elements.

In [9]:
A = {1,2,3,4,5,6,7}
B = {3,4,8,9}

Intersection = A & B #Another way: A.intersection(B)
Intersection

{3, 4}

# Problem 71. (Set Union)
Given two sets A and B, write a Python program to find their union and print all the distinct elements from both sets.

In [10]:
A = {1,2,3,4,5,6,7}
B = {3,4,8,9}

Union = A | B #Another way: A.union(B)
Union

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [11]:
A = {1,2,3,4,5,6,7}
B = {3,4,8,9}

Union = A.union(B)
Union

{1, 2, 3, 4, 5, 6, 7, 8, 9}

# Problem 72. (Set Difference) 
Given two sets A and B, write a Python program to find the difference between set A and set B (i.e., elements present in A but not in B) and print the result.

In [12]:
A = {1,2,3,4,5,6,7}
B = {3,4,8,9}

Difference = A - B  #Another way: A.difference(B)
Difference

{1, 2, 5, 6, 7}

# Problem 73: (Set Symmetric Difference:)
Given two sets A and B, write a Python program to find the symmetric difference between the two sets (i.e., elements that are present in either set A or set B, but not in both) and print the result.

In [15]:
set_A = {1,2,3,4,5,6,7}
set_B = {3,4,8,9}

Symmetric_Difference = set_A.symmetric_difference(set_B)
Symmetric_Difference

{1, 2, 5, 6, 7, 8, 9}

# Problem 74: (Set Operations)
Given three sets A, B, and C, write a Python program to find and print the intersection of A and B, the union of B and C, and the difference between C and A.

In [17]:
A = {1,2,3,4,5,6,7}
B = {3,4,8,9}
C = {1,2,10,12}

intersection_AB = A & B

union_BC = B | C

difference_CA = C - A

print('Intersection of A and B: ',intersection_AB)
print('Union of B and C: ',union_BC)
print('Difference between C and A: ',difference_CA)

Intersection of A and B:  {3, 4}
Union of B and C:  {1, 2, 3, 4, 8, 9, 10, 12}
Difference between C and A:  {10, 12}


# Problem 75: (Set Subset Check)
Given two sets A and B, write a Python program to check if set A is a subset of set B and print the result.

In [19]:
set_A = set(input("Enter elements for set A separated by spaces: ").split())
set_B = set(input("Enter elements for set B separated by spaces: ").split())

check_subset = set_A.issubset(set_B)

print("Set A:", set_A)
print("Set B:", set_B)

if check_subset:
    print("Set A is a subset of set B.")
else:
    print("Set A is not a subset of set B.")


Enter elements for set A separated by spaces: 1 2 3 4
Enter elements for set B separated by spaces: 1 3 4 6 8 9
Set A: {'3', '4', '2', '1'}
Set B: {'6', '4', '1', '8', '3', '9'}
Set A is not a subset of set B.


# Problem 76: (Set Superset Check)
Given two sets A and B, write a Python program to check if set A is a superset of set B and print the result.

In [20]:
def check_superset(A, B):
    if A.issuperset(B):
        print("Set A is a superset of set B")
    else:
        print("Set A is not a superset of set B")

set_A = {1, 2, 3, 4, 5}
set_B = {2, 4}

check_superset(set_A, set_B)


Set A is a superset of set B


# Problem 77: (Set Length Check)
Write a Python program that takes a set as input and prints the number of elements in the set.

In [21]:
set_A = set(input("Enter elements for set A separated by spaces: ").split())

set_size = len(set_A)

print('The number of elements in the set A is:',set_size )

Enter elements for set A separated by spaces: 1 2 3 
The number of elements in the set A is: 3


# Problem 76: (Set Membership Test)
Write a Python program that takes an element as input and checks if it exists in a given set. Print “Found” if the element is present and “Not Found” otherwise.

In [24]:
def check_element_in_set(element, input_set):
    if element in input_set:
        print(f"The element {element} exists in the set.")
    else:
        print(f"The element {element} does not exist in the set.")

example_set = {1, 2, 3, 4, 5}

element_to_check = int(input("Enter an element to check: "))

check_element_in_set(element_to_check, example_set)

Enter an element to check: 6
The element 6 does not exist in the set.


# Dictionary Problems: ( 77 - 86)

# Problem 77: (Dictionary Manipulation)
Given a dictionary with student names as keys and their corresponding scores as values, write a Python program to add a new student to the dictionary and update the score of an existing student.

In [1]:
dict_student = dict(radin=89,rafid=82,mohim=75)

print('Old dictionary: ',dict_student)

dict_student['rohin'] = 95

print('New dictionary: ',dict_student)

Old dictionary:  {'radin': 89, 'rafid': 82, 'mohim': 75}
Old dictionary:  {'radin': 89, 'rafid': 82, 'mohim': 75, 'rohin': 95}


# Problem 78: (Dictionary Keys and Values)
Write a Python program that takes a dictionary as input and prints all the keys and values in separate lines.

In [4]:
dict_student = dict(radin=89,rafid=82,mohim=75,rohin=95,rukaiya=90)

keys_list = dict_student.keys()
values_list = dict_student.values()

print(keys_list)
print(values_list)

dict_keys(['radin', 'rafid', 'mohim', 'rohin', 'rukaiya'])
dict_values([89, 82, 75, 95, 90])


# Problem 79: (Dictionary Length)
Write a Python program to calculate and print the number of key-value pairs in a given dictionary

In [5]:
dict_student = dict(radin=89,rafid=82,mohim=75,rohin=95,rukaiya=90)

print(f'Number of key-value pair of the dictionary is {len(dict_student)}.')

Number of key-value pair of the dictionary is 5.


# Problem 80: (Dictionary Value Search)
Given a dictionary of items and their prices, write a Python program to search for an item based on its price and print the item’s name.

In [7]:
#using for loop

items_prices = {'apple': 320,'banana': 100,'orange': 300,'grapes': 400,'mango': 200}

search_price = int(input("Enter the price to search: "))

found_item = None
for item, price in items_prices.items():
    if price == search_price:
        found_item = item
        break

if found_item:
    print(f"Item with price {search_price} found: {found_item}")
else:
    print(f"No item found with price {search_price}")


Enter the price to search: 400
Item with price 400 found: grapes


In [16]:
#using List Comprehension

items_prices = {'apple': 320,'banana': 100,'orange': 300,'grapes': 400,'mango': 200}

search_price = int(input("Enter the price to search: "))

found_items = [item for item, price in items_prices.items() if price == search_price]

if found_items:
    print(f"Item with price {search_price} found: {', '.join(found_items)}")
else:
    print(f"No item found with price {search_price}.")

Enter the price to search: 120
No item found with price 120.


# Problem 81: (Dictionary Merging)
Given two dictionaries, write a Python program to merge them into a single dictionary and print the result.

In [ ]:
#Using {**dict1, **dict2}

dict1 = {'a': 1, 'b': 2, 'c': 3}
dict2 = {'c': 4, 'd': 5, 'e': 6}

merged_dict2 = {**dict1, **dict2}

print("Merged Dictionary: ", merged_dict2)

# Problem 82: (Dictionary Key Removal)
Given a dictionary of items and their quantities, write a Python program to remove a specific item from the dictionary based on user input.

In [1]:
items_quantities = {
    'apple': 10,
    'banana': 5,
    'orange': 8,
    'grapes': 2,
    'mango': 4
}

user_input = input('Enter a item: ')

items_quantities.pop(user_input)

items_quantities

Enter a item: orange


{'apple': 10, 'banana': 5, 'grapes': 2, 'mango': 4}

# Problem 83. (Dictionary Sorting)
Given a dictionary with names as keys and corresponding ages as values, write a Python program to sort the dictionary based on age in ascending order.

In [1]:
dict_ages = {
    'Imran': 30,
    'Torik': 25,
    'Sohan': 35,
    'Rasel': 28,
    'Eva': 32
}

sorted_ages_dict = dict(sorted(dict_ages.items(), key=lambda x: x[1]))

#if we use x[0],the sort function works according to items not values.

print("Sorted Dictionary (Ascending Order):", sorted_ages_dict)


Sorted Dictionary (Ascending Order): {'Torik': 25, 'Rasel': 28, 'Imran': 30, 'Eva': 32, 'Sohan': 35}


# Problem 84: (Dictionary Frequency Count)
Write a Python program that takes a string as input and creates a dictionary containing each character as a key and its frequency as the value.

In [2]:
user_input = input("Enter a string: ")

char_frequency_dict = {}

for x in user_input:
    char_frequency_dict[x] = char_frequency_dict.get(x, 0) + 1

print(char_frequency_dict)

Enter a string: STUDYMART
{'S': 1, 'T': 2, 'U': 1, 'D': 1, 'Y': 1, 'M': 1, 'A': 1, 'R': 1}


# Problem 85: (Dictionary Comprehension)
Given a list of integers, write a Python program to create a dictionary where the keys are the elements from the list, and the values are their squares.

In [13]:
#using for loop

my_list = [1,2,5,4,6,8,7,9]
new_dict = {}

for x in my_list:
    new_dict[x] = x**2
    
print("Dictionary with Squares:")
print(new_dict)

Dictionary with Squares:
{1: 1, 2: 4, 5: 25, 4: 16, 6: 36, 8: 64, 7: 49, 9: 81}


In [9]:
#using a dictionary comprehension

my_list = [1,2,5,4,6,8,7,9]

new_dict = {num: num ** 2 for num in my_list}

print("Dictionary with Squares:")
print(new_dict)


Dictionary with Squares:
{1: 1, 2: 4, 5: 25, 4: 16, 6: 36, 8: 64, 7: 49, 9: 81}


In [12]:
#using lambda function

my_list = [1,2,5,4,6,8,7,9]

new_dict = dict(map(lambda x: (x, x ** 2), my_list))

print("Dictionary with Squares:")
print(new_dict)


Dictionary with Squares:
{1: 1, 2: 4, 5: 25, 4: 16, 6: 36, 8: 64, 7: 49, 9: 81}


# Program 86: (Dictionary Key Check)
Write a Python program that takes a key as input and checks if it exists in a given dictionary. Print “Key Found” if the key is present and “Key Not Found” otherwise.

In [19]:
input_key = input('Enter a key to check: ')

dict_student = dict(radin=89,rafid=82,mohim=75,rohin=95,rukaiya=90)

if input_key in dict_student:
    print('Key Found')
else:
    print('Key Not Found')

Enter a key to check: mizan
Key Not Found


In [17]:
input_key = input('Enter a key to check: ')

dict_student = dict(radin=89,rafid=82,mohim=75,rohin=95,rukaiya=90)

key_list = dict_student.keys()

for x in key_list:
    if x  == input_key:
        print('Key Found')
        break
    else:
        print('Key Not Found')
        break

Enter a key to check: roni
Key Not Found


# Nested Dictionary Problems: (87 - 91)

# Problem 87: (Access Nested Dictionary)
Given a nested dictionary containing student details, write a Python program to access and print specific information such as a student’s name, age, and address.

In [5]:
student_details = {'student1': {'name': 'Salman', 'age': 20, 'address': 'Wari, Dhaka',},
                   'student2': {'name': 'Sohan','age': 22, 'address': 'Jhiltuli, Faridpur',},
                   'student3': {'name': 'Imran','age': 21,'address': 'Boropul Rajbari',},}

student_key = 'student2'

if student_key in student_details:
    student_info = student_details[student_key]
    
    print(f"Student Name: {student_info['name']}")
    print(f"Student Age: {student_info['age']}")
    print(f"Student Address: {student_info['address']}")
    
else:
    print(f"Student with key '{student_key}' not found.")


Student Name: Sohan
Student Age: 22
Student Address: Jhiltuli, Faridpur


# Problem 88: (Nested Dictionary Length)
Write a Python program to calculate and print the total number of key-value pairs in a nested dictionary.

In [6]:
nested_dict = {
  'a': {'x': 1, 'y': 2},
  'b': {'z': 3, 'w': 4}
}

total_count = 0

for inner_dict in nested_dict.values():
  total_count += len(inner_dict)

print("The total number of key-value pairs in the nested dictionary is", total_count)


The total number of key-value pairs in the nested dictionary is 4


# Problem 89: (Nested Dictionary Update)
Given a nested dictionary of employee details, write a Python program to update an employee’s salary based on their employee ID.

In [1]:
def update_salary(employees, emp_id, new_salary):

  if emp_id in employees:
    
    employees[emp_id]['salary'] = new_salary
    
  else:
    print("Employee ID not found")
  
  return employees

employees = {
  '1001': {'name': 'Radin', 'salary': 50000},
  '1002': {'name': 'Rafid', 'salary': 60000},
  '1003': {'name': 'Nazmul', 'salary': 70000}
}

print("Original dictionary:")
print(employees)

employees = update_salary(employees, '1002', 65000)

print("Updated dictionary:")
print(employees)


Original dictionary:
{'1001': {'name': 'Radin', 'salary': 50000}, '1002': {'name': 'Rafid', 'salary': 60000}, '1003': {'name': 'Nazmul', 'salary': 70000}}
Updated dictionary:
{'1001': {'name': 'Radin', 'salary': 50000}, '1002': {'name': 'Rafid', 'salary': 65000}, '1003': {'name': 'Nazmul', 'salary': 70000}}


In [ ]:
employee_details = {
    'emp1': {
        'name': 'Radin',
        'position': 'Software Engineer',
        'salary': 60000,
    },
    'emp2': {
        'name': 'Roni',
        'position': 'Data Scientist',
        'salary': 70000,
    },
    'emp3': {
        'name': 'Rafid',
        'position': 'Project Manager',
        'salary': 80000,
    },
}


def update_salary(employee_id, new_salary):
    
    if employee_id in employee_details:
        
        employee_details[employee_id]['salary'] = new_salary
        
        print(f"Salary updated for {employee_id} is: {new_salary}")
    else:
        print(f"Employee ID '{employee_id}' not found.")


update_salary('emp2', 75000)

# Print updated employee details
print("Updated Employee Details:")
print(employee_details)


# Problem 90: (Nested Dictionary Sorting)
Given a nested dictionary containing product details (product name, price, and quantity), write a Python program to sort the products based on their prices in ascending order.

In [2]:
products = {
    "laptop": {"name": "Dell Inspiron", "price": 599.99, "quantity": 10},
    "smartphone": {"name": "Samsung Galaxy S21", "price": 799.99, "quantity": 20},
    "tablet": {"name": "Apple iPad", "price": 499.99, "quantity": 15},
    "earbuds": {"name": "Sony WF-1000XM4", "price": 279.99, "quantity": 25},
}

sorted_products = sorted(products.items(), key=lambda x: x[1]["price"])

sorted_products_dict = dict(sorted_products)

print("Sorted Product Details (Ascending Order by Price):")
for product_id, product_info in sorted_products_dict.items():
    print(f"{product_id}: {product_info}")

Sorted Product Details (Ascending Order by Price):
earbuds: {'name': 'Sony WF-1000XM4', 'price': 279.99, 'quantity': 25}
tablet: {'name': 'Apple iPad', 'price': 499.99, 'quantity': 15}
laptop: {'name': 'Dell Inspiron', 'price': 599.99, 'quantity': 10}
smartphone: {'name': 'Samsung Galaxy S21', 'price': 799.99, 'quantity': 20}


In [5]:
product_details = {
    'product1': {
        'name': 'Laptop',
        'price': 1200.00,
        'quantity': 5,
    },
    'product2': {
        'name': 'Smartphone',
        'price': 500.00,
        'quantity': 10,
    },
    'product3': {
        'name': 'Headphones',
        'price': 50.00,
        'quantity': 20,
    },
    'product4': {
        'name': 'Tablet',
        'price': 300.00,
        'quantity': 8,
    },
}

def sort_products_by_price(products):
        sorted_products = sorted(products.items(), key=lambda x: x[1]['price'])
        return dict(sorted_products)


sorted_product_details = sort_products_by_price(product_details)

print("Sorted Product Details (Ascending Order by Price):")
for product_id, product_info in sorted_product_details.items():
    print(f"{product_id}: {product_info}")


Sorted Product Details (Ascending Order by Price):
product3: {'name': 'Headphones', 'price': 50.0, 'quantity': 20}
product4: {'name': 'Tablet', 'price': 300.0, 'quantity': 8}
product2: {'name': 'Smartphone', 'price': 500.0, 'quantity': 10}
product1: {'name': 'Laptop', 'price': 1200.0, 'quantity': 5}


# Problem 91: (Nested Dictionary Key Search)
Write a Python program that takes a key as input and searches for it in a nested dictionary. If found, print the corresponding value, otherwise, print “Key Not Found.”

In [7]:
products = {
    "laptop": {"name": "Dell Inspiron", "price": 599.99, "quantity": 10},
    "smartphone": {"name": "Samsung Galaxy S21", "price": 799.99, "quantity": 20},
    "tablet": {"name": "Apple iPad", "price": 499.99, "quantity": 15},
    "earbuds": {"name": "Sony WF-1000XM4", "price": 279.99, "quantity": 25},
}


def find_key_recursive(d, key):
    if key in d: 
        return d[key]
    
    for k, v in d.items(): # Recursive case: iterate over the items
        
        if isinstance(v, dict): # Only recurse if the value is a dict
            
            value = find_key_recursive(v, key) # Call the function on the sub-dict
            
            if value: # If a value is returned, it means the key is found
                return value
    return None # If no value is returned, it means the key is not found


key = input("Enter a key to search: ")

result = find_key_recursive(products, key)

if result is None:
    print("Key Not Found.")
else:
    print(f"The value for key '{key}' is: {result}")


Enter a key to search: laptop
The value for key 'laptop' is: {'name': 'Dell Inspiron', 'price': 599.99, 'quantity': 10}


In [9]:
nested_dict = {
    'a': {
        'b': {
            'c': 1,
            'd': 2,
        },
        'e': {
            'f': 3,
            'g': 4,
        },
    },
    'h': {
        'i': {
            'j': 5,
            'k': 6,
        },
        'l': {
            'm': 7,
            'n': 8,
        },
    },
}


def search_key_in_nested_dict(dictionary, key):
    for k, v in dictionary.items(): #[here k=key, v=value in dict]
        if k == key:
            return v
        elif isinstance(v, dict): 
            result = search_key_in_nested_dict(v, key)
            if result is not None:
                return result
    return None

search_key = input("Enter the key to search: ")

result = search_key_in_nested_dict(nested_dict, search_key)

if result is not None:
    print(f"The value for key '{search_key}' is: {result}")
else:
    print(f"Key '{search_key}' not found.")


Enter the key to search: b
The value for key 'b' is: {'c': 1, 'd': 2}


# Break & Continue: (92 - 101)

# Problem 92: (Prime Number Checker)
Write a Python program that takes a number as input and determines if it is a prime number or not. Use a `for` loop to check for factors. If a factor is found, `break` out of the loop.

In [25]:
N = int(input('Enter the integer: '))

if N <= 1:
    print(f'{N} is not a prime number.')
    
elif N>1:
    for i in range(2,N):
       if N % i ==0:
        print(f'{N} is not a prime number.')
        print(f'{i} is a factor of {N}.')
        break
    else:
        print(f'{N} is a prime number.')
    

Enter the integer: 23
23 is a prime number.


# Problem 93 (Even Number Printer)
Write a Python program to print all even numbers from 1 to 20. Use a `for` loop and `continue` to skip odd numbers.

In [7]:
for number in range(1, 21):
    
    if number % 2 != 0:
        continue
        
    print(number)


2
4
6
8
10
12
14
16
18
20


# Problem 94: (Password Validator)
Write a Python program that takes a password as input and checks if it meets the following criteria: at least 8 characters long, contains both uppercase and lowercase letters, and has at least one digit. If the password is valid, print “Password accepted.” If not, use `continue` to prompt the user to enter a valid password.

In [2]:
while True:
    password = input("Enter a password: ")

    if len(password) < 8 or not any(char.isupper() for char in password) or not any(char.islower() for char in password) or not any(char.isdigit() for char in password):
        
        print("Invalid password. Please enter a password.")
        continue
        
    else:
        print("Password accepted.")
        break


Enter a password: jdthh
Invalid password. Please enter a password.
Enter a password: 8Fgjjjjj
Password accepted.


In [3]:
def validate_password(password):
  
  if len(password) < 8:
    return False

  if not any(char.isupper() for char in password):
    return False

  if not any(char.islower() for char in password):
    return False
  
  if not any(char.isdigit() for char in password):
    return False
  
  return True

while True:
  password = input("Enter a password: ")

  if validate_password(password):
    print("Password accepted.")
    break
  else:
    print("Enter a valid Passward.")


Enter a password: uhgy
Enter a valid Passward.
Enter a password: 5Lhhhhhh
Password accepted.


In [1]:
import re

passward = input('Enter the passward: ')

pass_len = len(passward)

is_valid = False

while True:
    if pass_len < 8: break
    
    elif not re.search('[A-Z]',passward): break
        
    elif not re.search('[a-z]',passward): break
        
    elif not re.search('[0-9]',passward): break
        
    elif re.search('\s',passward): break
    
    else:
        is_valid = True
        
if is_valid:
    print('Passward Accepted')

else:
    print('Enter a valid passward.')

Enter the passward: gjdhr
Enter a valid passward.


# Problem 95: (Divisible by 3 or 5)
Write a Python program to print all numbers from 1 to 50 that are divisible by either 3 or 5. Use a `for` loop and `continue` to skip numbers that are not divisible by either 3 or 5.

In [2]:
for number in range(1, 51):
    
    if number % 3 != 0 and number % 5 != 0:
        continue
    
    print(number)


3
5
6
9
10
12
15
18
20
21
24
25
27
30
33
35
36
39
40
42
45
48
50


# Problem 96: (Positive Number Sum)
Write a Python program that takes positive numbers as input until a negative number is entered. Then, calculate and print the sum of all positive numbers entered. Use a `while` loop and `break` to exit the loop when a negative number is encountered.

In [1]:
total_sum = 0

while True:
    number = float(input("Enter a positive number (or a negative number to exit): "))

    if number >= 0:
        total_sum += number
        
    else:
        break

print(f"The sum of positive numbers is: {total_sum}")


Enter a positive number (or a negative number to exit): 2
Enter a positive number (or a negative number to exit): 5
Enter a positive number (or a negative number to exit): 8
Enter a positive number (or a negative number to exit): -6
The sum of positive numbers is: 15.0


# Problem 97: (Word Palindrome Checker)
Write a Python program that takes a word as input and checks if it is a palindrome (reads the same forwards and backward). Use `continue` to skip checking the word if its length is less than 3 characters.

In [ ]:
s = input('Enter a string to check: ')
s = s.replace(" ", "").lower()

start = 0
end = len(s)-1
flag = True
while start < end:
    
    if len(s) < 3:
        continue
    
    if s[start] != s[end]:
        flag = False
    start+=1
    end-=1

if flag:
    print(f'Given string {s.upper()} is a palindrome.')
else:
    print(f'Given string {s.upper()} is not a palindrome.')

# Problem 98: (Odd Number Finder)
Write a Python program to find the first odd number from a list of integers. Use a `for` loop and `break` to stop the loop when the first odd number is found.

In [ ]:
int_list = [2, 4, 6, 8, 11, 12, 14, 16]

first_odd = None

for num in integer_list:
    
    if num % 2 != 0:
        first_odd = num
        break

if first_odd is not None:
    print(f"The first odd number in the list is: {first_odd}")
else:
    print("No odd number found in the list.")


# Problem 99: (Number Guessing Game)
Write a Python program that generates a random number between 1 and 100 and lets the user guess the number. Use a `while` loop, `break` when the correct number is guessed, and `continue` to keep prompting the user until they guess correctly.

In [14]:
import random
num = random.randint(1,20)
 
guess = int(input('Guess a num between 1 and 20'))

while guess != num:
    
    if guess < num:
        print("Your guess is too low.")
    
    elif guess > num:
        print("Your guess is too high.")
        
    guess = int(input("Guess again: "))

print('You guessed it right')

Guess a num between 1 and 2010
Your guess is too low.
Guess again: 12
Your guess is too low.
Guess again: 15
Your guess is too low.
Guess again: 18
Your guess is too low.
Guess again: 20
Your guess is too high.
Guess again: 19
You guessed it right


In [15]:
import random

secret_number = random.randint(1, 20)
    
attempts = 0

while True:
    guess = int(input("Enter your guess: "))
    attempts += 1
    
    if guess < secret_number:
        print("Too low! Try again.")
    elif guess > secret_number:
        print("Too high! Try again.")
    else:
        print(f"You guessed the correct number {secret_number} in {attempts} attempts.")
        break

Enter your guess: 15
Too high! Try again.
Enter your guess: 12
Too high! Try again.
Enter your guess: 9
Too high! Try again.
Enter your guess: 5
Too low! Try again.
Enter your guess: 7
Too high! Try again.
Enter your guess: 6
You guessed the correct number 6 in 6 attempts.


# Problem 100: (Vowel Counter)
Write a Python program that takes a string as input and counts the number of vowels (a, e, i, o, u) in it. Use a `for` loop and `continue` to skip counting non-vowel characters.

In [ ]:
input_string = input("Enter a string: ")

vowel_count = 0

for char in input_string:
  
    if char.lower() not in ['a', 'e', 'i', 'o', 'u']:
        continue
    
    vowel_count += 1

print(f"The number of vowels in the input string is: {vowel_count}")


# Problem 101: (Unique Characters) 
Write a Python program that takes a string as input and checks if it contains all unique characters (no character repeats). Use a `for` loop and `break` when a character repeats.

In [4]:
input_string = input("Enter a string: ")

unique_characters = set()

for char in input_string:
    
    if char in unique_characters:
        print("The string does not contain all unique characters.")
        break
    unique_characters.add(char)
else:
    print("The string contains all unique characters.")


Enter a string: love
The string contains all unique characters.


# Functions Problems: (102 -107)

# Problem 102: (Factorial Calculator)
Write a Python function called `factorial` that takes an integer as input and returns its factorial. Test the function with different values.

In [17]:
def factorial(n):
    if n < 0:
        return "Factorial is not defined for negative numbers."
    elif n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)

num = int(input('Enter a number: '))
result = factorial(num)
print(f"The factorial of {num} is: {result}")


Enter a number: 5
The factorial of 5 is: 120


# Problem 103: (Palindrome Checker)
Write a Python function called `is_palindrome` that takes a string as input and returns `True` if it is a palindrome and `False` otherwise. Test the function with different words.

In [27]:
def is_palindrome(s):
    s = s.replace(" ", "").lower()
    left, right = 0, len(s) - 1
    
    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
    
    return True


string = input("Enter a string to check palindrome: ")

is_palindrome(string)

Enter a string to check palindrome: madam


True

# Problem 104: (Even or Odd Checker) 
Write a Python function called `even_or_odd` that takes an integer as input and returns “Even” if the number is even and “Odd” if the number is odd. Test the function with different numbers.

In [28]:
def even_or_odd(a):
    if a % 2 ==0:
        return "Even"
    else:
        return "Odd"

num = int(input('Enter a number: '))
result = even_or_odd(num)
print(f"The number {num} is {result}.")

Enter a number: 7
The number 7 is Odd.


# Problem 105: (List Sum Calculator)
Write a Python function called `list_sum` that takes a list of integers as input and returns the sum of all elements in the list. Test the function with different lists.

In [1]:
def list_sum(input_list):
    total_sum = sum(input_list)
    return total_sum

list1 = [1, 2, 3, 4, 5]
list2 = [10, 20, 30, 40, 50]
list3 = [-1, -2, -3, -4, -5]

result1 = list_sum(list1)
result2 = list_sum(list2)
result3 = list_sum(list3)

print(f"Sum of list1: {result1}")
print(f"Sum of list2: {result2}")
print(f"Sum of list3: {result3}")


Sum of list1: 15
Sum of list2: 150
Sum of list3: -15


# Problem 106: (Greatest Common Divisor (GCD) Calculator)
Write a Python function called `gcd` that takes two integers as input and returns their greatest common divisor. Test the function with different pairs of numbers.

In [2]:
def gcd(a, b):
    while b:
        a, b = b, a % b
    return a

num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second number: "))

result = gcd(num1, num2)
print(f"The GCD of {num1} and {num2} is: {result}")

Enter the first number: 120
Enter the second number: 210
The GCD of 120 and 210 is: 30


# Problem 107: (Leap Year Checker)
Write a Python function called `is_leap_year` that takes a year as input and returns `True` if it is a leap year and `False` otherwise. Test the function with different years.

In [3]:
def is_leap_year(year):
    # Leap years are divisible by 4
    # However, years divisible by 100 are not leap years, unless they are also divisible by 400
    if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
        return True
    else:
        return False

year1 = 2020
year2 = 2021
year3 = 2000
year4 = 1900

print(f"Is {year1} a leap year? {is_leap_year(year1)}")
print(f"Is {year2} a leap year? {is_leap_year(year2)}")
print(f"Is {year3} a leap year? {is_leap_year(year3)}")
print(f"Is {year4} a leap year? {is_leap_year(year4)}")


Is 2020 a leap year? True
Is 2021 a leap year? False
Is 2000 a leap year? True
Is 1900 a leap year? False


# Nested Functions Problems: (108 - 110)

# Problem 108: (Math Operations)
Write a Python function called `math_operations` that takes three numbers and a string representing an operation (‘add’, ‘subtract’, ‘multiply’, or ‘divide’). The function should return the result of the specified operation on the three numbers. Implement the math operations as nested functions.

In [4]:
def math_operations(num1, num2, num3, operation):
    def add(x, y, z):
        return x + y + z

    def subtract(x, y, z):
        return x - y - z

    def multiply(x, y, z):
        return x * y * z

    def divide(x, y, z):
        # Check for division by zero
        if y == 0 or z == 0:
            return "Cannot divide by zero"
        return x / y / z

    # Perform the specified operation
    if operation == 'add':
        return add(num1, num2, num3)
    elif operation == 'subtract':
        return subtract(num1, num2, num3)
    elif operation == 'multiply':
        return multiply(num1, num2, num3)
    elif operation == 'divide':
        return divide(num1, num2, num3)
    else:
        return "Invalid operation"


result_add = math_operations(2, 3, 4, 'add')
result_subtract = math_operations(10, 5, 2, 'subtract')
result_multiply = math_operations(2, 3, 4, 'multiply')
result_divide = math_operations(10, 2, 2, 'divide')

print(f"Addition result: {result_add}")
print(f"Subtraction result: {result_subtract}")
print(f"Multiplication result: {result_multiply}")
print(f"Division result: {result_divide}")

Addition result: 9
Subtraction result: 3
Multiplication result: 24
Division result: 2.5


# Problem 109: (Greeting Generator)
Write a Python function called `greeting_generator` that takes a name as input and returns a greeting message using nested functions. The greeting message should be customizable (e.g., “Hello, {name}! How are you today?”).

In [7]:
def greeting_generator(name):
    def generate_greeting():
        return f"Hello, {name}!"

    def personalize_greeting():
        return f"{generate_greeting()} How are you today?"

    # Return the personalized greeting
    return personalize_greeting()

# Test the function with a name
name = input("Enter your name: ")
greeting_message = greeting_generator(name)
print(greeting_message)

Enter your name: Roni
Hello, Roni! How are you today?


# Problem 110: (Temperature Converter)
Write a Python function called `temperature_converter` that takes a temperature value and a string representing the scale (‘C’ for Celsius or ‘F’ for Fahrenheit) as input. The function should convert the temperature from one scale to the other using nested functions and return the converted value.

In [9]:
def temperature_converter(temperature, scale):
    def celsius_to_fahrenheit(celsius):
        return celsius * 9/5 + 32

    def fahrenheit_to_celsius(fahrenheit):
        return (fahrenheit - 32) * 5/9

    # Perform the temperature conversion based on the input scale
    if scale == 'C':
        converted_temperature = celsius_to_fahrenheit(temperature)
        return f"{temperature}°C is {converted_temperature}°F"
    elif scale == 'F':
        converted_temperature = fahrenheit_to_celsius(temperature)
        return f"{temperature}°F is {converted_temperature}°C"
    else:
        return "Invalid scale. Use 'C' for Celsius or 'F' for Fahrenheit."

# Test the function with a temperature and scale
temperature_value = float(input("Enter the temperature value: "))
temperature_scale = input("Enter the temperature scale (C/F): ")

result = temperature_converter(temperature_value, temperature_scale)
print(result)

Enter the temperature value: 100
Enter the temperature scale (C/F): C
100.0°C is 212.0°F


# Recursion Problems: (111 - 115)

# Problem 111. (Factorial Calculation with Recursion)
Write a recursive Python function called `factorial` that takes a non-negative integer as input and returns its factorial.

In [17]:
def factorial(n):
    if n < 0:
        return "Factorial is not defined for negative numbers."
    elif n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)

num = int(input('Enter a number: '))
result = factorial(num)
print(f"The factorial of {num} is: {result}")


Enter a number: 5
The factorial of 5 is: 120


# Problem 112: (Fibonacci Series) 
Write a recursive Python function called `Fibonacci` that takes an integer N as input and returns the Nth number in the Fibonacci series. The Fibonacci series is defined as follows: F(0) = 0, F(1) = 1, and F(n) = F(n-1) + F(n-2) for n > 1.

In [1]:
def Fibonacci(N):
    if N <= 0:
        return "Invalid input. Please provide a positive integer."
    elif N == 1:
        return 0
    elif N == 2:
        return 1
    else:
        return Fibonacci(N - 1) + Fibonacci(N - 2)

N = int(input("Enter a positive integer (N): "))
result = Fibonacci(N)

print(f"The {N}th number in the Fibonacci series is: {result}")

Enter a positive integer (N): 10
The 10th number in the Fibonacci series is: 34


# Problem 113: (The sum of Digits) 
Write a recursive Python function called `sum_of_digits` that takes an integer as input and returns the sum of its digits.

In [3]:
def sum_of_digits(number):
    if number < 10:
        return number
    else:
        # Recursive call to get the sum of digits for the remaining part of the number
        return number % 10 + sum_of_digits(number // 10)

integer_input = int(input("Enter an integer: "))
result = sum_of_digits(integer_input)

print(f"The sum of digits of {integer_input} is: {result}")


Enter an integer: 126
The sum of digits of 126 is: 9


# Power 115: (Power Calculation)
Write a recursive Python function called `power` that takes two positive integers, base and exponent, as input and returns the value of base raised to the exponent.

In [5]:
def power(base, exponent):
    
    if exponent == 0: # Base case: Any number raised to the power of 0 is 1
        return 1
    else:
        return base * power(base, exponent - 1)  # Recursive call to calculate base^exponent

base_input = int(input("Enter the base: "))
exponent_input = int(input("Enter the exponent: "))

result = power(base_input, exponent_input)

print(f"{base_input} raised to the power of {exponent_input} is: {result}")


Enter the base: 5
Enter the exponent: 3
5 raised to the power of 3 is: 125
